### PR TITLE
#2645 - Program deactivation UI - Part 2

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
#### Added a new actions menu for the programs available for Institutions and the ministry.

Institutions will see a label for "Edit program" or "View program" based in the fact that the program is editable or not.
Ministry will always see the label as "View program".

image

When a program is `Inactive` the badges are adjusted and the menu options are changed to "View program" and the `Deactivate` option is disabled.

image

Badges were also adjusted for the summary view, as shown below.

image

Program visualization adjusted to display the form is read-only when Inactive.
_An existing and known error was affecting the form.io to display the program in read-only.  The right visualization is read-only works depending in the order that the form.io definition and program data are loaded from the API. To fix it a main container was added to the form.io definition._

image

- API validation was added to prevent saving an inactive program.
- Created DB migration to allow the SABC code validation only for active programs. Following similar features, a unique index was used instead of a DB constraint to allow the `where` condition in DB validation.
- Minor CSS styles were created to allow the bold style to be applied to the menu action items.
- Created a `danger` global style for Vue following the same color as `Error`. The goal is to have a better context while setting the color.
- The same modal is shared for Ministry and Institution. The existing component was already executing client-type checks (they were kept, just refactored).

**Ministry Modal**
image

**Institution Modal**
image

### Modal Existing Issue
The latest developed modals introduced using the `keepModalOpen` option have an issue when the API call fails and the API call will not be executed again if the user retries the operation. This happens because the promise is not held once the API fails.
To resolve this issue and achieve the same behavior it was introduced the `beforeResolvePromise` callback that allow the modal to stay open and the retry operation to work. Please take a look and share some thoughts.

#### Bulk uploaded constraint adjustment
The bulk uploaded query to search programs was adjusted to include the `isActive` in the clause allowing the build upload to work also considering the deactivated programs.

### CODE REVIEW

**Review by lewischen-aot — APPROVED** [review]
Excellent work @andrewsignori-aot PR's looking really good 👍

File: sources/packages/web/src/components/common/modals/EducationProgramDeactivationModal.vue [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+          :rules="[checkNotesLengthRule]"
+          required
+          class="mt-4"
+          hide-details="auto"
+        ></v-textarea>
+      </template>
+      <template #footer>
+        <footer-buttons
+          primaryLabel="Confirm deactivation"
+          primaryButtonColor="danger"
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/common/modals/EducationProgramDeactivationModal.vue:33]: Was the primary color set to danger here highlighting the part that they are deactivating?

File: sources/packages/backend/apps/api/src/route-controllers/education-program/education-program.controller.service.ts [hunk 2]
```diff
@@ -71,6 +71,7 @@ export class EducationProgramControllerService {
         locationId: program.locationId,
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program/education-program.controller.service.ts:71]: Unused import in the file.

File: sources/packages/backend/apps/api/src/route-controllers/education-program/education-program.institutions.controller.ts [hunk 3]
```diff
@@ -94,7 +94,10 @@ export class EducationProgramInstitutionsController extends BaseController {
    * @param payload information to be updated.
    */
   @ApiUnprocessableEntityResponse({
-    description: "Not able to a save the program due to an invalid request.",
+    description:
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program/education-program.institutions.controller.ts:97]: Same error messages must be added here as well.

File: sources/packages/backend/apps/api/src/services/education-program/education-program.service.ts [hunk 4]
```diff
@@ -513,6 +521,7 @@ export class EducationProgramService extends RecordDataModelService<EducationPro
         "assessedBy.lastName",
         "programs.assessedDate",
         "programs.effectiveEndDate",
+        "programs.isActive",
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/services/education-program/education-program.service.ts:524]: Not part of the PR. The method at line 357 seems to be obsolete.

File: sources/packages/web/src/services/EducationProgramService.ts [hunk 5]
```diff
@@ -183,4 +184,16 @@ export class EducationProgramService {
...
   }
+
+  /**
+   * Allows a program to be deactivated.
+   * @param programId program to be deactivated.
+   * @param payload information to support the deactivation.
+   */
+  async deactivateProgram(
+    programId: number,
+    payload?: DeactivateProgramAPIInDTO,
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/EducationProgramService.ts:195]: Is there a reason to not prefer options here??

File: sources/packages/web/src/services/http/EducationProgramApi.ts [hunk 6]
```diff
@@ -165,4 +166,19 @@ export class EducationProgramApi extends HttpBaseClient {
...
   }
+
+  /**
+   * Allows a program to be deactivated.
+   * @param programId program to be deactivated.
+   * @param payload information to support the deactivation.
+   */
+  async deactivateProgram(
+    programId: number,
+    payload?: DeactivateProgramAPIInDTO,
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/http/EducationProgramApi.ts:177]: same here.

File: sources/packages/web/src/components/generic/FooterButtons.vue [hunk 7]
```diff
@@ -17,7 +17,7 @@
         class="ml-2"
         variant="elevated"
         data-cy="primaryFooterButton"
-        color="primary"
+        :color="primaryButtonColor"
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/generic/FooterButtons.vue:20]: 👍

File: sources/packages/web/src/components/common/modals/EducationProgramDeactivationModal.vue [hunk 8]
```diff
@@ -0,0 +1,92 @@
+<template>
+  <v-form ref="deactivateProgramForm">
+    <modal-dialog-base
+      :showDialog="showDialog"
+      title="Confirm deactivation"
+      ref="confirmDeactivationModal"
+      okLabel="Confirm deactivation"
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/common/modals/EducationProgramDeactivationModal.vue:7]: I am confused here on 2 things. the `okLabel` property belongs to `ConfirmModal` and not the `ModalDialogBase` right?? 

and I do not see that `keepModalOpen` is a property to the component, but an option to the resolvePromise method.

Let me know if I am missing something here.

File: sources/packages/web/src/components/common/modals/EducationProgramDeactivationModal.vue [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+    const submit = async () => {
+      const validationResult = await deactivateProgramForm.value.validate();
+      if (!validationResult.valid) {
+        return;
+      }
+      const payload = {
+        note: noteDescription.value,
+      };
+      await resolvePromise(payload);
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/common/modals/EducationProgramDeactivationModal.vue:70]: Should we await on resolvePromise to reset the form?? Is it to allow the user to submit the same information again when the promise is rejected??

File: sources/packages/web/src/components/common/ProgramDetails.vue [hunk 10]
```diff
@@ -110,50 +154,81 @@ export default defineComponent({
...
+    ): Promise<boolean> => {
+      if (modalResult === false) {
+        return true;
+      }
+      try {
+        await ApiClient.EducationProgram.deactivateProgram(
+          props.programId,
+          modalResult as DeactivateProgramAPIInDTO,
+        );
+        snackBar.success("Program deactivate with success.");
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/common/ProgramDetails.vue:215]: deactivated ?

File: sources/packages/web/src/components/common/ProgramDetails.vue [hunk 11]
```diff
@@ -79,22 +104,41 @@
       />
     </v-col>
   </v-row>
+  <education-program-deactivation-modal
+    ref="declineEducationProgramModal"
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/common/ProgramDetails.vue:108]: Ref value seems to be wrong (declineEducationProgramModal)

**Review by dheepak-aot — COMMENTED** [review]
Great and extensive work. Thanks for handling the duplicate SABC code part. Please have a look at the comments.

File: sources/packages/forms/src/form-definitions/educationprogram.json [hunk 12]
```diff
@@ -9,1221 +9,395 @@
...
-            "labelMargin": "",
-            "tag": "h2",
-            "className": "category-header-large primary-color",
-            "attrs": [
+            "title": "Education Program Main Container",
+            "customClass": "formio-panel-unset",
+            "collapsible": false,
+            "hideLabel": true,
+            "key": "educationProgramMainContainer",
+            "logic": [
```
Comment (dheepak-aot) [inline sources/packages/forms/src/form-definitions/educationprogram.json:17]: 👍

File: sources/packages/web/src/components/common/modals/EducationProgramDeactivationModal.vue (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/modals/EducationProgramDeactivationModal.vue:33]: Yes, I belive that the intention was to be extra clear about the action impact because there is no way to revert it.

File: sources/packages/backend/apps/api/src/route-controllers/education-program/education-program.institutions.controller.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/education-program/education-program.institutions.controller.ts:97]: Added the "duplicate SABC code." to the message since the other ones are related only to the update.
Please let me know if that was the expectation.

File: sources/packages/backend/apps/api/src/services/education-program/education-program.service.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/services/education-program/education-program.service.ts:524]: Removing it 👍

File: sources/packages/web/src/components/common/ProgramDetails.vue [hunk 13]
```diff
@@ -4,19 +4,44 @@
...
+        </template>
+        <v-list
+          active-class="active-list-item"
+          density="compact"
+          bg-color="default"
+          color="primary"
+        >
+          <v-list-item @click="goToProgram" :title="programActionLabel" />
+          <v-divider-inset-opaque />
+          <check-permission-role
```
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/ProgramDetails.vue:31]: Same here, and it also has a nice comment 😄

File: sources/packages/web/src/components/common/ProgramDetails.vue (hunk 11 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/ProgramDetails.vue:108]: Can you please clarify what is wrong exactly?

File: sources/packages/web/src/components/common/modals/EducationProgramDeactivationModal.vue (hunk 9 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/modals/EducationProgramDeactivationModal.vue:70]: Yes, waiting will keep the form's current state and allow it to be submitted again.

File: sources/packages/web/src/services/EducationProgramService.ts (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/EducationProgramService.ts:195]: No extra properties needed or planned so I kept it as it is right now.
Is the idea that any optional property will always be present inside an options parameter?

File: sources/packages/web/src/components/common/ProgramDetails.vue [hunk 14]
```diff
@@ -110,50 +154,81 @@ export default defineComponent({
...
+      return "Edit";
     });
 
-    const programButtonAction = () => {
-      if (isInstitutionUser.value) {
-        router.push({
+    const goToProgram = () => {
+      if (AuthService.shared.authClientType === ClientIdType.Institution) {
+        return router.push({
           name: InstitutionRoutesConst.EDIT_LOCATION_PROGRAMS,
```
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/ProgramDetails.vue:177]: Good point, we already have a `VIEW_LOCATION_PROGRAMS` defined as below:
```
location-programs/location/:locationId/program-view/:programId
```
It points to the below page that shows the program summary and its location offerings.
image

Right now I would not change it.

File: sources/packages/backend/apps/api/src/route-controllers/education-program/education-program.controller.service.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/education-program/education-program.controller.service.ts:70]: Removed.

File: sources/packages/backend/apps/api/src/services/education-program/education-program.service.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/services/education-program/education-program.service.ts:482]: Method removed, I did not find any reference to it.

File: sources/packages/web/src/components/common/modals/EducationProgramDeactivationModal.vue (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/modals/EducationProgramDeactivationModal.vue]: Properties removed.

File: sources/packages/web/src/components/common/ProgramDetails.vue (hunk 10 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/ProgramDetails.vue]: Fixed.

File: sources/packages/web/src/components/common/ProgramDetails.vue (hunk 11 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/ProgramDetails.vue:108]: Got it now, renamed to `deactivateEducationProgramModal`.

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot

File: sources/packages/web/src/services/EducationProgramService.ts (hunk 5 above)
Comment (dheepak-aot) [reply sources/packages/web/src/services/EducationProgramService.ts:195]: Not necessarily, keeping it inside options make it easy to extend the optional properties as we know. Im good here.

**Review by dheepak-aot — APPROVED** [review]
Thanks for making the changes. 👍Looks good.
